File: python/countAndSay.py
def countAndSay(n: int) -> str:
    return recurse(n)


# Time -> 0(N*M) Space -> 0(N + M)
def recurse(n: int) -> str:
    if n == 1:
        return "1"

    rle = recurse(n - 1)

    idx = 0
    builder = []
    while idx < len(rle):
        currIdx = idx
        while idx < len(rle) and rle[currIdx] == rle[idx]:
            idx += 1

        count = f"{idx - currIdx}"
        builder.append(count)
        builder.append(rle[currIdx])

    return "".join(builder)


# Time -> 0(N*M) Space -> 0(M)
def approach1(n: int) -> str:
    rle = "1"
    i = 1
    while i < n:
        builder = []
        idx = 0
        while idx < len(rle):
            currIdx = idx
            while idx < len(rle) and rle[currIdx] == rle[idx]:
                idx += 1

            count = f"{idx - currIdx}"
            builder.append(count)
            builder.append(rle[currIdx])
        rle = "".join(builder)
        i += 1
    return rle


print(countAndSay(10))
print(approach1(10))

File: go/countAndSay/countAndSay.go
package main

import (
	"fmt"
	"strings"
)

func countAndSay(n int) string {
	return recurse(n)
}

// Time -> 0(N*M) Space -> 0(N + M)
func recurse(n int) string {
	if n == 1 {
		return "1"
	}

	rle := recurse(n - 1)

	builder := &strings.Builder{}

	idx := 0
	for idx < len(rle) {
		currIdx := idx
		for idx < len(rle) && rle[currIdx] == rle[idx] {
			idx++
		}
		length := fmt.Sprintf("%d", (idx - currIdx))
		builder.WriteString(length)
		builder.WriteByte(rle[currIdx])
	}

	return builder.String()
}

// Time -> 0(N*M) Space -> 0(M)
func approach1(n int) string {
	rle := "1"
	for i := 1; i < n; i++ {
		idx := 0
		builder := &strings.Builder{}
		for idx < len(rle) {
			currIdx := idx
			for idx < len(rle) && rle[currIdx] == rle[idx] {
				idx++
			}
			length := fmt.Sprintf("%d", (idx - currIdx))
			builder.WriteString(length)
			builder.WriteByte(rle[currIdx])
		}

		rle = builder.String()
	}

	return rle
}

func main() {
	fmt.Println(countAndSay(10))
	fmt.Println(approach1(10))
}
